### PR TITLE
Kernel: Make processes start with a 16-byte-aligned stack

### LINKED ISSUE #6785: Crash with SSE2 because stack is not always aligned to 16-byte boundary.
Currently, as of commit 3c0355a398b7b7677da116f0a1e75116b58aeaf9 , when you build with SSE2 SystemServer will crash due to a general protection fault when trying to run `String::vformatted()` very early into boot. It appears that it is caused by an attempt to call an sse2 instruction (`15908:       0f 29 85 38 ff ff ff    movaps %xmm0,-0xc8(%ebp)`). This causes an exception since at the time that it attempts to call this instruction, the stack is not 16 byte aligned (%ebp=0x0250004c). GCC assumes that stack frames are 16-byte aligned since about GCC 4.7 IIRC, and it appears that somewhere along the line we are violating this assumption, since %ebp = 0xc (mod 16). I've pasted the stack trace below.

```
[NetworkTask(7:7)]: NetworkTask: E1000NetworkAdapter network adapter found: hw=52:54:00:12:34:56
[SystemServer(6:6)]: CRASH: CPU #0 General protection fault in ring 3       
[#0 SystemServer(6:6)]: Exception code: 0000 (isr: 0000)                                                                                
[#0 SystemServer(6:6)]:     pc=001b:0a65c908 eflags=00010216                                                                            
[#0 SystemServer(6:6)]:  stack=0023:024fff6c                                                                                            
[#0 SystemServer(6:6)]:     ds=0023 es=0023 fs=0023 gs=002b                                                                             
[#0 SystemServer(6:6)]:    eax=0250009c ebx=0a6b8fa8 ecx=0250009c edx=0000000b
[#0 SystemServer(6:6)]:    ebp=0250004c esp=c1c6af88 esi=024fffa8 edi=0250007c
[#0 SystemServer(6:6)]:    cr0=80010013 cr2=c1b45c00 cr3=01291000 cr4=00340ee4
[#0 SystemServer(6:6)]: 0x0a65c908  (?)                                                                                                 
                                                                                                                                        
[#0 SystemServer(6:6)]: 0xc0372577  Kernel::Process::crash(int, unsigned int, bool) +0x333
[#0 SystemServer(6:6)]: 0xc056c4a0  illegal_instruction_handler +0x0                                                                    
[#0 SystemServer(6:6)]: 0xc056c7b5  page_fault_handler +0x0                                                                             
[#0 SystemServer(6:6)]: 0xc0563734  general_protection_fault_asm_entry +0x26
[#0 SystemServer(6:6)]: Process regions:                                                                                                
[#0 SystemServer(6:6)]: BEGIN       END         SIZE        ACCESS  NAME    
[#0 SystemServer(6:6)]: 02401000 -- 02500fff 00100000 RW  T  Stack (Main thread)
[#0 SystemServer(6:6)]: 02502000 -- 02502fff 00001000 R XS C Signal trampoline
[#0 SystemServer(6:6)]: 02510000 -- 0251ffff 00010000        malloc: ChunkedBlock(256)
[#0 SystemServer(6:6)]: 02530000 -- 0253ffff 00010000 RW     malloc: ChunkedBlock(64)
[#0 SystemServer(6:6)]: 02550000 -- 0255ffff 00010000 RW     malloc: ChunkedBlock(32)
[#0 SystemServer(6:6)]: 0a647000 -- 0a6b7fff 00071000 R XS C /usr/lib/Loader.so    
[#0 SystemServer(6:6)]: 0a6b8000 -- 0a6b9fff 00002000 RW     /usr/lib/Loader.so (data-rw)
[#0 SystemServer(6:6)]: Kernel regions:                                                                                                 
[#0 SystemServer(6:6)]: BEGIN       END         SIZE        ACCESS  NAME           
[#0 SystemServer(6:6)]: c1001000 -- c1100fff 00100000 RW     kmalloc subheap       
[#0 SystemServer(6:6)]: c1104000 -- c1104fff 00001000 RW     HPET MMIO             
[#0 SystemServer(6:6)]: c1106000 -- c1106fff 00001000 R    C Signal trampolines             
[#0 SystemServer(6:6)]: c1108000 -- c1117fff 00010000 RW  T  Kernel stack (thread 0)
[#0 SystemServer(6:6)]: c1119000 -- c1128fff 00010000 RW  T  Kernel stack (thread 1)
[#0 SystemServer(6:6)]: c113b000 -- c114afff 00010000 RW  T  Kernel stack (thread 3)    
[#0 SystemServer(6:6)]: c114c000 -- c115bfff 00010000 RW  T  Kernel stack (thread 4)    
[#0 SystemServer(6:6)]: c115d000 -- c115dfff 00001000  W     UHCI Framelist         
[#0 SystemServer(6:6)]: c115f000 -- c1160fff 00002000  W     UHCI Queue Head Pool       
[#0 SystemServer(6:6)]: c1162000 -- c1163fff 00002000  W     UHCI Transfer Descriptor Pool
[#0 SystemServer(6:6)]: c1165000 -- c1174fff 00010000 RW  T  Kernel stack (thread 5)
[#0 SystemServer(6:6)]: c1176000 -- c1179fff 00004000 RW     VirtIO MMIO         
[#0 SystemServer(6:6)]: c117b000 -- c117bfff 00001000 RW     VirtIO Queue
[#0 SystemServer(6:6)]: c117d000 -- c117dfff 00001000 RW     VirtIO Queue                 
[#0 SystemServer(6:6)]: c117f000 -- c117ffff 00001000 RW     VirtIOConsole Receive
[#0 SystemServer(6:6)]: c1181000 -- c1181fff 00001000 RW     VirtIOConsole Transmit
[#0 SystemServer(6:6)]: c1183000 -- c1186fff 00004000 RW     VirtIO MMIO
[#0 SystemServer(6:6)]: c1188000 -- c1188fff 00001000 RW     VirtIO Queue
[#0 SystemServer(6:6)]: c118a000 -- c118afff 00001000 RW     VirtIORNG
[#0 SystemServer(6:6)]: c118c000 -- c118cfff 00001000 RW     E1000 RX
[#0 SystemServer(6:6)]: c118e000 -- c118efff 00001000 RW     E1000 TX
[#0 SystemServer(6:6)]: c1190000 -- c11affff 00020000 RW     E1000 MMIO
[#0 SystemServer(6:6)]: c11b1000 -- c11b2fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11b4000 -- c11b5fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11b7000 -- c11b8fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11ba000 -- c11bbfff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11bd000 -- c11befff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11c0000 -- c11c1fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11c3000 -- c11c4fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11c6000 -- c11c7fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11c9000 -- c11cafff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11cc000 -- c11cdfff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11cf000 -- c11d0fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11d2000 -- c11d3fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11d5000 -- c11d6fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11d8000 -- c11d9fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11db000 -- c11dcfff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11de000 -- c11dffff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11e1000 -- c11e2fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11e4000 -- c11e5fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11e7000 -- c11e8fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11ea000 -- c11ebfff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11ed000 -- c11eefff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11f0000 -- c11f1fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11f3000 -- c11f4fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11f6000 -- c11f7fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11f9000 -- c11fafff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11fc000 -- c11fdfff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c11ff000 -- c1200fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c1202000 -- c1203fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c1205000 -- c1206fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c1208000 -- c1209fff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c120b000 -- c120cfff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c120e000 -- c120ffff 00002000 RW     E1000 RX buffer
[#0 SystemServer(6:6)]: c1211000 -- c1212fff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c1214000 -- c1215fff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c1217000 -- c1218fff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c121a000 -- c121bfff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c121d000 -- c121efff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c1220000 -- c1221fff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c1223000 -- c1224fff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c1226000 -- c1227fff 00002000 RW     E1000 TX buffer
[#0 SystemServer(6:6)]: c1229000 -- c1229fff 00001000 RW     IDE PRDT
[#0 SystemServer(6:6)]: c122b000 -- c122bfff 00001000 RW     IDE DMA region
[#0 SystemServer(6:6)]: c122d000 -- c122dfff 00001000 RW     IDE PRDT
[#0 SystemServer(6:6)]: c122f000 -- c122ffff 00001000 RW     IDE DMA region
[#0 SystemServer(6:6)]: c1231000 -- c1232fff 00002000 RW     AHCI HBA
[#0 SystemServer(6:6)]: c1234000 -- c1234fff 00001000 RW     AHCI Port Command List
[#0 SystemServer(6:6)]: c1236000 -- c1236fff 00001000 RW     AHCI Port Command List
[#0 SystemServer(6:6)]: c1238000 -- c1238fff 00001000 RW     AHCI Port Command List
[#0 SystemServer(6:6)]: c123a000 -- c123afff 00001000 RW     AHCI Port Command List
[#0 SystemServer(6:6)]: c123c000 -- c123cfff 00001000 RW     AHCI Port Command List
[#0 SystemServer(6:6)]: c123e000 -- c123efff 00001000 RW     AHCI Port Command List
[#0 SystemServer(6:6)]: c1240000 -- c1240fff 00001000 RW     Ext2FS: Block group descriptors
[#0 SystemServer(6:6)]: c1242000 -- c1c05fff 009c4000 RW     KBuffer
[#0 SystemServer(6:6)]: c1c07000 -- c1c55fff 0004f000 RW     KBuffer
[#0 SystemServer(6:6)]: c1c57000 -- c1c57fff 00001000 RW     Ext2FS: Cached bitmap block
[#0 SystemServer(6:6)]: c1c59000 -- c1c59fff 00001000 RW     Ext2FS: Cached bitmap block
[#0 SystemServer(6:6)]: c1c5b000 -- c1c6afff 00010000 RW  T  Kernel stack (thread 6)
[#0 SystemServer(6:6)]: c1c6c000 -- c1c6cfff 00001000 RW     Ext2FS: Cached bitmap block
[#0 SystemServer(6:6)]: c1c6e000 -- c1c6efff 00001000 RW     Ext2FS: Cached bitmap block
[#0 SystemServer(6:6)]: c1c70000 -- c1c7ffff 00010000 RW  T  Kernel stack (thread 7)
[#0 SystemServer(6:6)]: c1c81000 -- c1c90fff 00010000 RW     Kernel Packet Buffer
[#0 FinalizerTask(4:4)]: Generating coredump for pid: 6
[FinalizerTask(4:4)]: RTC: Year: 2021, month: 4, day: 30, hour: 16, minute: 30, second: 24
[#0 FinalizerTask(4:4)]: Can't find directory '/tmp/coredump' for core dump
```

### PR DESCRIPTION
**Pull request by ccapitalK**
Fixes #6785.

### CODE REVIEW

**Comment by ccapitalK** [comment]
https://research.csiro.au/tsblog/debugging-stories-stack-alignment-matters/ Is a fairly good explanation of the stack alignment requirements of spawned processes.

File: Kernel/Syscalls/execve.cpp [hunk 1]
```diff
@@ -131,6 +131,9 @@ static KResultOr<FlatPtr> make_userspace_stack_for_main_thread(Region& region, V
 
     // NOTE: The stack needs to be 16-byte aligned.
     new_esp -= new_esp % 16;
+    // GCC assumes that the return address has been pushed to the stack when it enters the function,
+    // so we need to reserve an extra 4 bytes to make GCC's stack alignment calculations work
+    new_esp -= 4;
```
Comment (IdanHo) [inline Kernel/Syscalls/execve.cpp:136]: How about `sizeof(void*)` to make this compatible with x86_64 as well?

Comment (ccapitalK) [reply Kernel/Syscalls/execve.cpp:136]: I think the ABI alignment restrictions might be different on x86_64, let me check.

Comment (tomuta) [reply Kernel/Syscalls/execve.cpp:136]: Also, can you add a `VERIFY` that will catch this going forward?

Comment (ccapitalK) [reply Kernel/Syscalls/execve.cpp:136]: Oh yeah, a VERIFY would probably be good. Will add that shortly.

Comment (ccapitalK) [reply Kernel/Syscalls/execve.cpp:136]: @IdanHo Apparently it does work the same way on x86 (1 (ptr size) below boundary), I'll update it now. Also, we should eventually move this to an arch specific directory.

**Pull request merged**